modules/graceful/manager.go

```diff
@@ -288,6 +318,42 @@ func (g *gracefulManager) setState(st state) {
 
 // InformCleanup tells the cleanup wait group that we have either taken a listener
 // or will not be taking a listener
-func (g *gracefulManager) InformCleanup() {
+func (g *Manager) InformCleanup() {
 	g.createServerWaitGroup.Done()
 }
+
+// Done allows the manager to be viewed as a context.Context, it returns a channel that is closed when the server is finished terminating
+func (g *Manager) Done() <-chan struct{} {
+	g.lock.RLock()
+	if g.done == nil {
+		g.lock.RUnlock()
+		g.lock.Lock()
+		if g.done == nil {
+			g.done = make(chan struct{})
+		}
+		defer g.lock.Unlock()
+		return g.done
+	}
+	defer g.lock.RUnlock()
+	return g.done
```

Comment:
This RLock/Lock/RUnlock/Unlock ordering seems weird. What would happen if we do....?:
```golang
g.lock.RLock()
defer g.lock.RUnlock()
if g.done == nil {
	g.lock.Lock()
	defer g.lock.Unlock()
	if g.done == nil {
		g.done = make(chan struct{})
	}
}
return g.done
```

Reply:
You can't get a Lock whilst holding an RLock.

Reply:
I mean we could just make it write lock but that sort of misses the benefit of the read lock. It's just unfortunate that you can escalate locks in go. 

Another option is to have explicit separate read and write locks but that's risking locking write without locking read.

Reply:
We had this conversation before, sorry... It just takes me off guard that the sync library is so primitive. 😮

Unless you see a fundamental flaw, I think it's much easier to just initialize `g.done` at the start. Much easier to follow, easier to maintain, no weird locking...

Reply:
I've changed to initialise all four channels at the start.

Reply:
It looks cleaner